CONTRIBUTING.md

```diff
@@ -1,91 +1,379 @@
+
+2. We allow some exceptions: e.g. allowing using !, ? for Akka to send actor message.
+
+Before submitting a PR, you should always run style check first:
+```
+  ## Run style check for compile, test, and integration test.
+  sbt scalastyle test:scalastyle it:scalastyle
+```
+
+<a name="write-unittest"></a>
+## How to write unit test
+TBD
+
+<a name="write-integrationtest"></a>
+## How to write integration test
+TBD
+
+<a name="write-doc"></a>
+## How to write document
+Documentation contributions are very welcome!
+
+You can contribute documentation by pull request, same as code contribution.
+Main directory is ```docs/```, and you can refer to docs/README.md for how to build / test documentation.
+
+NOTE: this documentation is the release documentation for Gearpump, not the website documentation at Apache website, which is maintained by another repository.
+
+
+<a name="committer-work"></a>
+# Committer section
+_This section applies to committers only._
+
+<a name="approve-pull-request"></a>
+## Approve pull request
+It's committer's duty to review pull requests from contributors. 
+
+Any PR ready to merge shall have at least one +1(s) and no -1(s) from other than the one who authored this PR. And any merge shall wait another 24 hours after the first +1 received to wait for potential comments.
+Only committer has the right to perform PR merge to Apache upstream.
+
+
+<a name="merge-pull-request"></a>
+## Merge pull request
+For each committer, it is suggested to make a local clone of Apache Gearpump repository and make it solely for PR merge purpose. 
+```bash
+git clone https://git-wip-us.apache.org/repos/asf/incubator-gearpump
+
+```
+**NOTE: Only Apache Gearpump repository is writable. The GitHub mirror is READONLY.**
+
+
+With this local copy of Apache official repo at hand, we are now able to perform merge for PRs. 
+
+**Important: A pull request must first be properly approved before you are allowed to merge it.**
+
+To pull in a merge request you should generally follow the command line instructions sent out by GitHub.
+
+1. Go to your local copy of the [Apache git repo](https://git-wip-us.apache.org/repos/asf/incubator-gearpump.git), switch
+   to the `master` branch, and make sure it is up to date.
+
+        $ git checkout master
+        $ git fetch origin
+        $ git merge origin/master
+
+2. Create a local branch for integrating and testing the pull request.  You may want to name the branch according to the
+  Gearpump JIRA ticket associated with the pull request (example: `GEARPUMP-1234`).
+
+        $ git checkout -b <local_test_branch>  # e.g. git checkout -b GEARPUMP-1234
+
+3. Merge the pull request into your local test branch.
+
+        $ git pull <remote_repo_url> <remote_branch>
+    You can use `./dev-tools/gearpump-merge.py <pull-number>` to produce the above command most of the time.
+
+4.  Assuming that the pull request merges without any conflicts:
```

Comment:
I'm afraid an extra rebase step is needed to avoid an extra merge commit in case master is updated although having no conflicts with the branch.